gdnative-derive/src/variant/from.rs

```diff
@@ -61,46 +54,55 @@ pub(crate) fn expand_from_variant(derive_data: DeriveData) -> Result<TokenStream
 
             let var_input_ident_iter = std::iter::repeat(&var_input_ident);
 
+            // Return `FromVariantError` if input is an uninhabitable enum
+            let early_return = variants.is_empty().then(|| {
+                quote! {
+                    return Err(FVE::UnknownEnumVariant {
+                        variant: __key,
+                        expected: &[],
+                    });
+                }
+            });
+
             quote! {
-                {
-                    let __dict = ::gdnative::core_types::Dictionary::from_variant(#input_ident)
-                        .map_err(|__err| FVE::InvalidEnumRepr {
-                            expected: VariantEnumRepr::ExternallyTagged,
-                            error: std::boxed::Box::new(__err),
-                        })?;
-
-                    let __keys = __dict.keys();
-                    if __keys.len() != 1 {
-                        Err(FVE::InvalidEnumRepr {
-                            expected: VariantEnumRepr::ExternallyTagged,
-                            error: std::boxed::Box::new(FVE::InvalidLength {
-                                expected: 1,
-                                len: __keys.len() as usize,
-                            }),
-                        })
-                    }
-                    else {
-                        let __key = String::from_variant(&__keys.get(0))
-                            .map_err(|__err| FVE::InvalidEnumRepr {
-                                expected: VariantEnumRepr::ExternallyTagged,
-                                error: std::boxed::Box::new(__err),
-                            })?;
-                        match __key.as_str() {
-                            #(
-                                #ref_var_ident_string_literals => {
-                                    let #var_input_ident_iter = &__dict.get_or_nil(&__keys.get(0));
-                                    (#var_from_variants).map_err(|err| FVE::InvalidEnumVariant {
-                                        variant: #ref_var_ident_string_literals,
-                                        error: std::boxed::Box::new(err),
-                                    })
-                                },
-                            )*
-                            variant => Err(FVE::UnknownEnumVariant {
-                                variant: variant.to_string(),
-                                expected: &[#(#ref_var_ident_string_literals),*],
-                            }),
-                        }
-                    }
```

Comment:
This `{}` block is no longer needed?

Reply:
Yeah, it's almost identical to L68 ~ L105. (expected the `early_return` statement)

There are so many lines of change due to indentation. The outermost `{}` pair is removed because `return_expr` is the only expression used inside generated `from_variant` implementation. Should I keep it?

https://github.com/godot-rust/godot-rust/blob/94d7a45672825c916bfd5425b1703a0db1a822a5/gdnative-derive/src/variant/from.rs#L68-L105

Reply:
I see, thanks. No it's fine, you can leave your changes like this 🙂